### DISCUSSION TITLE
请问下CommConnEntry中list的作用是什么？

### DISCUSSION THREAD

**Discussion by sourcema** [sogou/workflow#1384] [Q&A]
这个链表作用是什么？

**Comment by sourcema** [comment]
对于server，每个client的entry中的target中的idle_list能够穿起来所有的idle的client的connection？我看实现每个client对应的target都是new出来的，没太懂这个idle_list的作用？

**Comment by holmes1412** [comment]
`struct CommConnEntry`是用于记录连接上下文的，无论是server里的accept_conn()还是client里的launch_conn()，都需要malloc一份出来。

这个entry可能会被挂到多种链表上，比如空闲链表、keepalive链表等。

它内部的`list`成员，就是用来把entry挂到链表上的，链表使用的是kernel的list结构。

一般我们自己实现的链表可能是：
```
struct list_node {
    void *entry;
    struct list_node *next;
};
```
那我们每次要挂到链表上的时候，分配好了entry之后，还得分配一个list_node，才能挂起来。

但是kernel的实现利用entry自带的list作为内部拉链，所以这个list，就可以让操作链表的时候少分配一次list_node。

**Reply by sourcema** [reply]
链表的结构我能理解，我不太懂的是：对于server的每个client的CommConnEntry中的CommTarget记录的每个client的地址信息对吧，其中idle_list保存的是什么？所有client对应的空闲连接链表？没太看懂这块，还请解答下。

**Reply by holmes1412** [reply]
server和client是类似的用法，只不过对于server来说，一个对方target的idle_list上只会有一个等待被回复的entry

**Reply by sourcema** [reply]
那client上的target是可以有多个server么?还有我看CommScheduler.h中CommScheduler::request里*target = object->acquire(wait_timeout);
client中的target是通过它获取的，这个是干嘛的呢？

**Reply by holmes1412** [reply]
- 应该说一个client的一个target对应了一个server，那么可能有多个连接，上下文可能被放到idle_list上，这样在类似于keepalive的场景，就不用每次都新建连接了；
- acquire是要拿一个可用的target，从而对这个target进行请求；

**Reply by sourcema** [reply]
好的，谢谢，我再研究下
